core/store/src/flat/delta.rs

```diff
@@ -0,0 +1,204 @@
+use borsh::{BorshDeserialize, BorshSerialize};
+
+use near_primitives::state::ValueRef;
+use near_primitives::types::{RawStateChangesWithTrieKey, ShardId};
+use std::collections::HashMap;
+
+#[cfg(feature = "protocol_feature_flat_state")]
+use super::store_helper;
+use crate::{CryptoHash, StoreUpdate};
+
+#[derive(BorshSerialize, BorshDeserialize)]
+pub struct KeyForFlatStateDelta {
+    pub shard_id: ShardId,
+    pub block_hash: CryptoHash,
+}
+
+/// Delta of the state for some shard and block, stores mapping from keys to value refs or None, if key was removed in
+/// this block.
+#[derive(BorshSerialize, BorshDeserialize, Clone, Default, Debug, PartialEq, Eq)]
+pub struct FlatStateDelta(pub(crate) HashMap<Vec<u8>, Option<ValueRef>>);
+
+impl<const N: usize> From<[(Vec<u8>, Option<ValueRef>); N]> for FlatStateDelta {
+    fn from(arr: [(Vec<u8>, Option<ValueRef>); N]) -> Self {
+        Self(HashMap::from(arr))
+    }
+}
+
+impl FlatStateDelta {
+    /// Assumed number of bytes used to store an entry in the cache.
+    ///
+    /// Based on 36 bytes for `ValueRef` + guessed overhead of 24 bytes for `Vec` and `HashMap`.
+    pub(crate) const PER_ENTRY_OVERHEAD: u64 = 60;
+
+    /// Returns `Some(Option<ValueRef>)` from delta for the given key. If key is not present, returns None.
+    pub fn get(&self, key: &[u8]) -> Option<Option<ValueRef>> {
```

Comment:
API nit: is there a reason this is returning the field by-value? It this returned a reference, the user could very well clone for themselves if they actually need to do so.